### PR TITLE
[Feature]: Add SerialExtract, Optimize Vector Index

### PR DESCRIPTION
**Pull request by arjunsk**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [X] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue # https://github.com/matrixorigin/matrixone/issues/14064#issuecomment-1884443572

## What this PR does / why we need it:

- Implementing serial_extract function to extract individual values from serialized columns.
- Optimized Vector Index Mapping to utilize the `min` with `serial_full`.

## Examples

```mysql
mysql> select serial_extract(pair, 0 as varchar(3)), serial_extract(pair, 1 as bigint) from(select serial_full(cast("aby" as varchar(3)), cast(1 as bigint)) as pair);
+---------------------------------------+-----------------------------------+
| serial_extract(pair, 0 as varchar(3)) | serial_extract(pair, 1 as bigint) |
+---------------------------------------+-----------------------------------+
| aby                                   |                                 1 |
+---------------------------------------+-----------------------------------+

-- For NULL datatype doesn't matter.
mysql> select serial_extract(pair, 0 as varchar(3)), serial_extract(pair, 1 as bigint) from(select serial_full(cast(NULL as varchar(3)), cast(1 as bigint)) as pair);
+---------------------------------------+-----------------------------------+
| serial_extract(pair, 0 as varchar(3)) | serial_extract(pair, 1 as bigint) |
+---------------------------------------+-----------------------------------+
| NULL                                  |                                 1 |
+---------------------------------------+-----------------------------------+
```

### CODE REVIEW

File: pkg/container/types/tuple.go [hunk 1]
```diff
@@ -747,6 +747,10 @@ func decodeTuple(b []byte) (Tuple, int, error) {
 		case b[i] == stringTypeCode:
 			el, off = decodeBytes(b[i+1:])
 			off += 1
+		case b[i] == enumCode:
+			//TODO: need to verify @YANGGMM
+			el, off = decodeUint(uint16Code, b[i+1:])
```
Comment (arjunsk) [inline pkg/container/types/tuple.go:752]: @YANGGMM is this the correct decode impl for Enum?

File: pkg/sql/parsers/tree/expr.go [hunk 2]
```diff
@@ -1121,6 +1121,56 @@ type ResolvableTypeReference interface {
...
+	ctx.WriteString(" as ")
+	node.ResultType.(*T).InternalType.Format(ctx)
+	ctx.WriteByte(')')
+}
+
+// Accept implements NodeChecker interface
+func (node *SerialExtractExpr) Accept(v Visitor) (Expr, bool) {
+	//TODO: need validation from @iamlinjunhong
+
+	newNode, skipChildren := v.Enter(node)
```
Comment (arjunsk) [inline pkg/sql/parsers/tree/expr.go:1145]: @iamlinjunhong Please validate if this implementation is correct or not.

File: pkg/sql/plan/function/list_builtIn.go [hunk 3]
```diff
@@ -1272,6 +1272,35 @@ var supportedStringBuiltIns = []FuncNew{
 		},
 	},
 
+	// function `serial_extract`
+	{
+		functionId: SERIAL_EXTRACT,
+		class:      plan.Function_STRICT | plan.Function_ZONEMAPPABLE,
+		//TODO: verify with @m-schen
+		//layout:     STANDARD_FUNCTION,
+		layout: CAST_EXPRESSION,
```
Comment (arjunsk) [inline pkg/sql/plan/function/list_builtIn.go:1281]: @m-schen Is it ok to use CAST_EXPRESSION here? Or should I use STANDARD_FUNCTION?

**Comment by arjunsk** [comment]
Adding `serial_min` and optimization for `vector index` in the PR.

File: pkg/container/types/tuple.go (hunk 1 above)
Comment (YANGGMM) [reply pkg/container/types/tuple.go:774]: It should be fine.

**Review by XuPeng-SH — APPROVED** [review]

**Review by ouyuanning — APPROVED** [review]

File: pkg/sql/plan/function/list_builtIn.go (hunk 3 above)
Comment (m-schen) [reply pkg/sql/plan/function/list_builtIn.go:1281]: the layout field just effect how the function shows in the layout of `explain`.

standard_function is func(arg0, arg1, arg2...)

and the cast is cast(arg0 as type).
I think it must not meet your requirement. you can write a new one for you if the standard is not what you want. @arjunsk

Comment (arjunsk) [reply pkg/sql/plan/function/list_builtIn.go]: Thank you for the suggestion. I will use `STANDARD_FUNCTION`.

**Review by iamlinjunhong — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Review by m-schen — APPROVED** [review]

**Review by YANGGMM — APPROVED** [review]

**Review by aressu1985 — APPROVED** [review]

**Pull request merged**